### DISCUSSION TITLE
PrivateRoute redirecting when Authenticated

### DISCUSSION THREAD

**Discussion by HunterJuneau** [nss-evening-cohort-14/evening-client-side#322] [Capstones!] [answered]
Hunter Juneau

On initial load, `user` in my App is showing false, which is redirecting me to my unauthenticated page. But then my user updates to normal, but I am still stuck on my unauthenticated page. Trying to go back to the rest of the app just repeats the cycle.

Here is my auth code in App:

```
const [user, setUser] = useState(false);

  useEffect(() => {
    firebase.auth().onAuthStateChanged((authed) => {
      if (authed) {
        console.warn(authed);
        const userInfoObj = {
          fullName: authed.displayName,
          uid: authed.uid,
        };
        setUser(userInfoObj);
      } else {
        setUser(false);
      }
    });
  }, []);
```

and here is my PrivateRoute code: 

```
const PrivateRoute = ({ component: Component, user, ...rest }) => {
  const routeChecker = (taco) => (
    user ? (
      <Component {...taco} uid={user.uid} />
    ) : (
      <Redirect to={{ pathname: '/unauthenticated', state: { from: taco.location } }} />
    )
  );
  return <Route {...rest} render={(props) => routeChecker(props)} />;
};

PrivateRoute.propTypes = {
  component: PropTypes.func,
  user: PropTypes.any,
};
```

[My current code](https://github.com/HunterJuneau/achievement-tracker/tree/create-achievements)

**Comment by drteresavasquez** [comment]
@HunterJuneau try to modify user on initial load to be null or something other than false and let us know what happens

**Reply by HunterJuneau** [reply]
That caused the error `Cannot read property 'uid' of null` on these useEffect's in my Router:
```
export default function Routes({ user }) {
  const [achievements, setAchievements] = useState([]);
  const [games, setGames] = useState([]);

  useEffect(() => getAchievements(user.uid).then(setAchievements), []);

  useEffect(() => getGames(user.uid).then(setGames), []);
```

**Reply by HunterJuneau** [reply]
I moved my useEffects out of Router which fixed the errors in my comment above, but I'm still having the issue in my original ticket.

**Reply by drteresavasquez** [reply]
@HunterJuneau as we talked about, move all of your data to the way we have done it in class and you should be able to follow the same stuff we did in all the projects we have completed. See the sample code and watch some of the videos over this.

**Reply by drteresavasquez** [reply]
Closing until you do what is prescribed and let us know if you need something else

**Reply by HunterJuneau** [reply]
> Closing until you do what is prescribed and let us know if you need something else

I have moved all of my data and useEffect's out of my router, but it seems to have had no effect on my issue. Here's my current code: https://github.com/HunterJuneau/achievement-tracker/tree/create-achievements

**Reply by drteresavasquez** [reply]
Did you set up your app the way we have done it in class? Also, make sure that you have a flow chart for your data and post it here.

**Reply by HunterJuneau** [reply]
> 
> 
> Did you set up your app the way we have done it in class? Also, make sure that you have a flow chart for your data and post it here.

I believe so, it looks the same anyway.

[Data Flow Chart](https://docs.google.com/presentation/d/1UKe2_nfhDuoGah0M4lmr4b_Qb7N80aM_G37YldTiirQ/edit?usp=sharing)

**Reply by drteresavasquez** [reply]
OK. Try this. Pass the user to each of your components. You are passing it to the routes but the components may need it as well. There is a more elegant way to do this that we can show you after our 1:1s tonight, but I think that will handle your view issues.

**Reply by drteresavasquez** [reply]
So as an example, you have this, see comment inside of sample.
```javascript
 <PrivateRoute
      exact
      path='/'
      user={user}
      component={() => (
        <Achievements
           // PASS USER HERE user={user}
           achievements={achievements}
           setAchievements={setAchievements}
           setGames={setGames}
         />
      )}
  />
```

**Reply by drteresavasquez** [reply]
And can you share your flow chart to anyone with the link?

**Reply by HunterJuneau** [reply]
Tried what you suggested, still no change.

Flow chart is fixed.

**Reply by drteresavasquez** [reply]
Go through and look at your views. You are not using user. You are using `uid` so this is something that you are going to have to work through your data flow and make sure all the views and components are getting what they need.

Also, use the tools you have. React dev tools should be the first thing you look at for state and props.

**Reply by HunterJuneau** [reply]
I am passing uid, but these Private Routes aren't being loaded in the first place. I am always redirected to my Unauthenticated page. In my React dev tools, user is showing up correctly.

**Reply by drteresavasquez** [reply]
Try handling the user update in the unauthenticated view...if the user is updated, redirect them

**Reply by HunterJuneau** [reply]
Do you mean adding something like this to my Unauthenticated View?

```
useEffect(() => {
    <Redirect to='/' />;
  }, [user]);
```

**Reply by HunterJuneau** [reply]
Did the above and no change.

**Comment by drteresavasquez** [comment, marked as answer]
Figured out @HunterJuneau was using the `NavLink` instead of `Link` so the app was refreshing causing issues.
